dask/array/overlap.py

```diff
@@ -23,6 +24,119 @@
 from .creation import empty_like, full_like
 
 
+class ArrayOverlapLayer(Layer):
+    """Simple HighLevelGraph array overlap layer.
+
+    Lazily computed High-level graph layer for a array overlap operations.
+
+    Parameters
+    ----------
+    name : str
+        Name of new output overlap array.
+    array : Dask array
+    axes: Mapping
+        Axes dictionary indicating overlap in each dimension,
+        e.g. ``{'0': 1, '1': 1}``
+    """
+
+    name: str
+    # array: # Dask array
+    axes: Mapping[str, int]
+
+    def __init__(
+        self,
+        name,
+        axes,
+        chunks,
+        numblocks,
+        token,
+    ):
+        self.name = name
+        self.axes = axes
+        self.chunks = chunks
+        self.numblocks = numblocks
+        self.token = token
+        self._cached_keys = None
```

Comment:
I feel like this is an ugly way to do things - we're passing in a whole bunch of stuff in the initialization that is really only being used in the _construct_graph method. The aim is to get things working and then tidy it up, but I want to point this out now regardless.

There's also some potential confusion with all the different sorts of `name` (there is a name equivalent to the original `x.chunks`, a name that is `"overlap-"+tokenize(x, axes)`, and then a `getitem` name as well). Ideally this would be simplified.

Reply:
> I feel like this is an ugly way to do things - we're passing in a whole bunch of stuff in the initialization that is really only being used in the _construct_graph method. The aim is to get things working and then tidy it up, but I want to point this out now regardless.

Yeah, that's a fair concern -- I would say that it comes down to the fact that we are sort-of reimplementing some of the logic belonging to a dask collection, but in a more restricted setting (no deserialization, no numpy/pandas), so it's both (a) kind of duplicative feeling, and (b) a little awkward. Perhaps @rjzamora has some ideas about patterns for making things like this less verbose.

> There's also some potential confusion with all the different sorts of `name` (there is a name equivalent to the original `x.chunks`, a name that is `"overlap-"+tokenize(x, axes)`, and then a `getitem` name as well). Ideally this would be simplified.

Yep, I agree! I'm certainly open to ideas for how to name things in a clearer way (so to speak).

At the very least, I think some of these things could be computed in `__init__` rather than passed in. This would help it be more self-consistent as well. For example, `numblocks` is computable from `chunks`, so they aren't really independent parameters to be passed in. You may also be able to `tokenize` (x.name, axes) so you wouldn't have to pass in a `token` (I'm not actually sure if there would be other consequences to changing how `getitem`-name is computed)